### ISSUE TITLE
'gif diff' changes scrollbar position

### ISSUE THREAD

**Issue by jkloth** [Microsoft/WSL#627]
When using 'git diff' on a large enough diff that is several pages in size, the scrollbar position is moved with the position in the view.  Just scrolling down is not an issue, but it becomes apparent when then scrolling back up (say to the top).  If you then exit (press 'q') when at the top, the screen buffer position stays at the scrolled position instead of moving to the end of the screen buffer.

This then becomes a problem when running 'git diff' again as it overwrites the screen buffer producing a garbled display.  The regular 'less' command does not show this behavior.  It may be a by-product of the coloring that git uses in its output.

A workaround for this as a user is to always advance to the end of the diff (press '>') before exiting ('q').

**Comment by fpqc** [comment]
jkloth are you using an X terminal or PuTTY through ssh, or are you using the native windows console?  It sounds like the console, so I'll give you my boilerplate regarding console problems.

The native Windows console interface for WSL has way too many problems to deal with (unicode problems, weird character glitches, lines getting jumbled, etc), and the devs here have said that a major upgrade to the Windows console (features and bugfixes) probably won't make it into the July 30th release of Win 10 Anniversary edition (Redstone), which is their current deadline for getting the rest of WSL in order.  It is therefore highly recommended to use a terminal emulator (rxvt-unicode-256color is highly recommended) connected to a Windows X server (Xming is very good) or do some workarounds to get sshd running and then connecting through PuTTY.

**Comment by zadjii-msft** [comment]
@jkloth Which build are you on? A little while back I changed how the console handled full-window applications so that it should behave much nicer. I want to make sure you have that update because I think it should fix the issue you're describing.

Generally, the windows support for complex commandline applications is pretty good, however, the console's code moves a little slower than the kernel's code, which makes bug fixes come out a bit slower than the kernel can add and fix things.

**Comment by jkloth** [comment]
At the time of the comment, it was build 14372(?), but the issue still presents itself on 14379.

Now applications like nano and less work as I would expect, but whatever git uses for its own pager makes things a mess.

**Comment by jkloth** [comment]
After some investigation, 'git diff' pipes its output to less with the -F -R -X options.  The interesting one is '-X' which prevents less from sending the termcap initialization and deinitialization strings (to prevent clearing the screen on exit presumably).  And THEN less uses either termcap 'insert_line' (al) or 'scroll_reverse' (sr) to do the scrolling.  It chooses which to use based on which one is shorter AND if 'memory_above' (da) is set.

Since WSL is using 'xterm' as its TERM, al = \e[L, sr = \eM, da = false, resulting in scroll_reverse being used and the scrollbar being moved (as it should given the termcap).

In short, the console is doing what it is told, HOWEVER, it seems that another termcap might be in order.  Or to "break" the terminal operations to match how other terminal handle this.

I'm currently digging through the terminfo database to see if any of the definitions match the behavior (and other capabilities).

**Comment by jkloth** [comment]
More information.  After digging through the terminfo database, it seems none of the relevant descriptions match the console's behavior.  With a thorough reading (and re-reading, and re-reading...) of the terminfo(5) manpage, it seems that without its own terminfo entry, the behavior should (must?) be changed.

Quoting:
> The boolean **ndstr** should be set if each scrolling window is effectively a
> view port on a screen-sized canvas.  To test for this capability, create a scrolling region in the middle
> of  the  screen, write  something to the bottom line, move the cursor to the top of the region, and do
> **ri** followed by **dl1** or **ind**.  If the data scrolled off the bottom of the region by  the  **ri**
> reappears,  then  scrolling  is  non-destructive.   System  V  and  XSI  Curses  expect  that  **ind**,
>  **ri**, **indn**, and **rin** will simulate destructive scrolling; their documentation cautions you not to
> define **csr** unless this is true.  This curses implementation is more liberal and will do explicit 
> erases after scrolling if **ndstr** is defined.

Given that, console needs to handle the **ri** and **ind** operations as if there was no screen buffer.

**Comment by zadjii-msft** [comment]
@jkloth 

My bad that I lost track of this. Any chance this is still happening for you? I know it's been 6 months and there have been quite a few changes to the conhost codebase since then, but I did just solve a trio of bugs that may have played into this behavior and I was curious.  Let me know.

**Comment by bitcrazed** [comment]
Thanks for the discussion.

Closing this issue since:
1. This issue is very old and has likely been fixed in recent Win10 builds
1. This is the WSL issues repo, but this is an issue in Console which has its own [Console GitHub Repo](https://github.com/Microsoft/console)
1. GitHub doesn't allow issues to be moved between repos, preserving posters' identity :(

If you have further asks/issues, please file new issues on our [Console GitHub Repo](https://github.com/Microsoft/console).
